src/3devoUpdate/Form1.cs

```diff
@@ -1,610 +1,541 @@
+          if( !dfuSeFileValidation.FirmwareFileValidation(usbDevice, "3devo-AiridMainboard", dfuUtilCmdLine.dfuseAddress) ) {
+            EnableInterface(false /*No upload button*/);
+            return;
+          }
+
+          dfuUtilCmdLine.Generate();
+          dfuUtil.Launch(dfuUtilCmdLine.command, DfuUtil.CommandType.FILE_UPLOAD);
+          break;
+        case USBDeviceList.MachineType.None:
+        default:
+          throw new Exception("StartUploadProcess(); No uploadable device selected.");
+          break;
+      }
+    }
+    #endregion
+
+    #region UI Events
+
+    // Drag and drop
+    private void Event_DragEnter( object sender, DragEventArgs e ) {
+      if( e.Data.GetDataPresent(DataFormats.FileDrop) )
+        e.Effect = DragDropEffects.Copy;
+    }
+
+    // Drag and drop
+    private void Event_DragDrop( object sender, DragEventArgs e ) {
+      string[] files = (string[])e.Data.GetData(DataFormats.FileDrop);
+      if( ((GroupBox)sender).Name == "gbFlashFile" )
+        txtFlashFile.Text = files[0];
+    }
 
-        // Upload!
-        private void btnUpload_Click(object sender, EventArgs e)
-        {
-            selected_port_name = this.port;
-            avrdude.launch(this.cmdBox);     
+    // Browse for flash file
+    private void BtnFlashBrowse_Click( object sender, EventArgs e ) {
+      downloadIsReady = false; // Something has changed in the selection, so let's check if everything is ok later on.
+      btnUpload.Enabled = false;
+
+      try {
+        //openFileDialog1.Filter += "|All files (*.*)|*.*"; // Probably not a necessary option.
+        switch( DeviceInfoList.ElementAt(cmbPort.SelectedIndex).MachineName ) {
+          case USBDeviceList.MachineType.FilamentMaker:
+            openFileDialog1.Title = "Select Filemant Maker or Filament Extruder firmware";
+            openFileDialog1.Filter = "Hex files (*.hex)|*.hex";
+            break;
+          case USBDeviceList.MachineType.StBootloader:
+          case USBDeviceList.MachineType.AiridDryer:
+            openFileDialog1.Title = "Select Airid Dryer firmware";
+            openFileDialog1.Filter = "Binary files (*.DfuSe)|*.DfuSe";
+            break;
+          case USBDeviceList.MachineType.None:
+          default:
+            throw new Exception("BtnFlashBrowse_Click(); This model is not yet supported.");
+            break;
         }
-                
-        // Resize console when form resizes
-        private void Form1_Resize(object sender, EventArgs e)
-        {
-            //txtConsole.Height = Height - txtConsole.Top - 64;
+      }
+      catch( ArgumentOutOfRangeException ) {
+        Console.WriteLine("BtnFlashBrowse_Click(); ArgumentOutOfRangeException: Pressed while no combobox items are available.");
+        return;
+      }
+
+      DialogResult Result = DialogResult.None;
+      Result = openFileDialog1.ShowDialog();
+
+      if( Result == DialogResult.OK ) {

```

Comment:
Wouldn't it be better to do dfuse validation here, rather than when pressing upload?

Reply:
In this case not really, I think. The file validation needs the device vid, pid, etc as parameters, meaning that the validation check would need to run every time another device is selected with the combobox as well. That would also generate a new message prompt when the newly selected device doens't match the file as device type (even when mistakenly selecting the wrong device). Also from user perspective I would rather have just a single error message (if any) when the selection process is finished and upload is pressed :-)

Reply:
Good point about changing the selected device, I was mostly thinking of a linear "select device" -> "select file" -> "upload".

> Also from user perspective I would rather have just a single error message (if any) when the selection process is finished and upload is pressed :-)

I don't agree there - When you select an inappropriate file, that seems to be the right moment to provide this feedback.

One alternative that could solve this, would be to not provide feedback using a popup, but add a text label underneath the file selection or so that displays the error. So if you select an inappropriate or invalid file, the "Upload" button would be disabled and the error shown in the label. If you then select a different device that makes the file inappropriate, the error message is still shown, without being annoying.